Detectors/ZDC/calib/src/InterCalib.cxx

```diff
@@ -0,0 +1,386 @@
+  int ih = -1;
+  if (hn.EqualTo("hZNA")) {
+    ih = hidZNA;
+  } else if (hn.EqualTo("hZPA")) {
+    ih = hidZPA;
+  } else if (hn.EqualTo("hZNC")) {
+    ih = hidZNC;
+  } else if (hn.EqualTo("hZPC")) {
+    ih = hidZPC;
+  } else if (hn.EqualTo("hZEM")) {
+    ih = hidZEM;
+  } else {
+    printf("Not recognized histogram name: %s\n", hname);
+    return -1;
+  }
+  clear(ih);
+  const int32_t dim = 6;
+  double x[dim];
+  int32_t bins[dim];
+  int64_t nb = hs->GetNbins();
+  int64_t nn = 0;
+  printf("Histogram %s has %ld bins\n", hname, nb);
+  double cutl = mInterCalibConfig->cutLow[ih];
+  double cuth = mInterCalibConfig->cutHigh[ih];
+  double contt = 0;
+  for (int64_t i = 0; i < nb; i++) {
+    double cont = hs->GetBinContent(i, bins);
+    if (cont <= 0) {
+      continue;
+    }
+    for (int32_t d = 0; d < dim; ++d) {
+      x[d] = hs->GetAxis(d)->GetBinCenter(bins[d]);
+    }
+    if (TMath::Nint(x[5] - ic) == 0 && x[0] > cutl && x[0] < cuth) {
+      nn++;
+      contt += cont;
+      cumulate(ih, x[0], x[1], x[2], x[3], x[4], cont);
+    }
+  }
+  int ierr = mini(ih);
+  if (h_corr[ih]) {
+    delete h_corr[ih];
+    h_corr[ih] = nullptr;
+  }
+  if (hc_corr[ih]) {
+    delete hc_corr[ih];
+    hc_corr[ih] = nullptr;
+  }
+  if (ierr) {
+    printf("FAILED processing RUN2 data for %s ih = %d\n", hname, ih);
+  } else {
+    printf("Processed RUN2 data for %s ih = %d\n", hname, ih);
+    replay(ih, hs, ic);
+  }
+  printf("Trigger class selection %d and %d bins %g events and cuts (%g:%g): %ld\n", ic, nn, contt, cutl, cuth);
+  return 0;
+}
+
+void InterCalib::replay(int ih, THnSparse* hs, int ic)
+{
+  auto* cfg = mInterCalibConfig;
+  // clang-format off
+  if(ih == 0)h_corr[ih] = new TH1F("hZNASc","ZNA sum corr",cfg->nb1[ih],cfg->amin1[ih],cfg->amax1[ih]);
+  if(ih == 1)h_corr[ih] = new TH1F("hZPASc","ZPA sum corr",cfg->nb1[ih],cfg->amin1[ih],cfg->amax1[ih]);
+  if(ih == 2)h_corr[ih] = new TH1F("hZNCSc","ZNC sum corr",cfg->nb1[ih],cfg->amin1[ih],cfg->amax1[ih]);
+  if(ih == 3)h_corr[ih] = new TH1F("hZPCSc","ZPC sum corr",cfg->nb1[ih],cfg->amin1[ih],cfg->amax1[ih]);
+  if(ih == 4)h_corr[ih] = new TH1F("hZEM2c","ZEM2"   ,cfg->nb1[ih],cfg->amin1[ih],cfg->amax1[ih]);
+  if(ih == 0)hc_corr[ih] = new TH2F("cZNAc","ZNA;TC;SUM corr",cfg->nb2[ih],cfg->amin2[ih],cfg->amax2[ih],cfg->nb2[ih],cfg->amin2[ih],cfg->amax2[ih]);
+  if(ih == 1)hc_corr[ih] = new TH2F("cZPAc","ZPA;TC;SUM corr",cfg->nb2[ih],cfg->amin2[ih],cfg->amax2[ih],cfg->nb2[ih],cfg->amin2[ih],cfg->amax2[ih]);
+  if(ih == 2)hc_corr[ih] = new TH2F("cZNCc","ZNC;TC;SUM corr",cfg->nb2[ih],cfg->amin2[ih],cfg->amax2[ih],cfg->nb2[ih],cfg->amin2[ih],cfg->amax2[ih]);
+  if(ih == 3)hc_corr[ih] = new TH2F("cZPCc","ZPC;TC;SUM corr",cfg->nb2[ih],cfg->amin2[ih],cfg->amax2[ih],cfg->nb2[ih],cfg->amin2[ih],cfg->amax2[ih]);
+  if(ih == 4)hc_corr[ih] = new TH2F("cZEMc","ZEM;ZEM1;ZEM2 corr",cfg->nb2[ih],cfg->amin2[ih],cfg->amax2[ih],cfg->nb2[ih],cfg->amin2[ih],cfg->amax2[ih]);
```

Comment:
This method can run just on RUN2 data because it uses the THnSparse. For RUN3 we will likely need to make a test reconstruction and verify calibration or alternatively design some caching of processed data. We need to discuss the framework in more detail.